### ISSUE TITLE
Objectness loss using iou of prediction and gt boxes?

### ISSUE THREAD

**Issue by sreddy-es** [ultralytics/yolov5#6952]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Thank you for the great work. I was looking at how the the target of objectness loss is calculated. In models/utils/loss.py. I was looking at compute_loss()

For bbox loss
```
pbox = torch.cat((pxy, pwh), 1)  # predicted box
iou = bbox_iou(pbox.T, tbox[i], x1y1x2y2=False, CIoU=True)  # iou(prediction, target)
lbox += (1.0 - iou).mean()  # iou loss
```

For objectness loss the exact iou which is later used for bbox loss is used to compute objectness loss targets.
```
# Objectness
score_iou = iou.detach().clamp(0).type(tobj.dtype)
if self.sort_obj_iou:
    sort_id = torch.argsort(score_iou)
    b, a, gj, gi, score_iou = b[sort_id], a[sort_id], gj[sort_id], gi[sort_id], score_iou[sort_id]
tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * score_iou  # iou ratio
```
I had a look at [Gaussian YOLOv3](https://arxiv.org/pdf/1904.04620.pdf) on your suggestion in #1863. Where it states that we compute iou of the most suitable anchor and the gt bbox which should be a part of the bbox loss. I don't understand why it's included in objectness loss in this case, because I would think that we'd need the iou or overlap between the grid cell and and gt bbox to get the objectness target values. 

Please let me know if I'm missing something.

**Comment by glenn-jocher** [comment]
@sreddy-es that's correct, YOLOv5 assigns  anchor-target prediction IoU as the objectness target. This means that YOLOv5 objectness outputs are accuracy predictors, which helps sort low-accuracy detections during NMS.

**Comment by sreddy-es** [comment]
@glenn-jocher can you clarify what anchor-target prediction IoU means? Is it IoU between target-prediction or target-anchor? My issue is that the it looks like the code is using IoU between target-prediction as the objectness target. Is that intentional? if so then you've lost me because Gaussian YOLO v3 is doing something different.

**Comment by glenn-jocher** [comment]
@sreddy-es just IoU between the prediction (aka anchor) and the label (aka target)

**Comment by sreddy-es** [comment]
Thank you @glenn-jocher for the prompt replies!

**Comment by talentCODE** [comment]
> @sreddy-es that's correct, YOLOv5 assigns anchor-target prediction IoU as the objectness target. This means that YOLOv5 objectness outputs are accuracy predictors, which helps sort low-accuracy detections during NMS.

yolov3 set  1 as the  target of objectness  and the author of yolov3 said it would be better than use the IoU as target.   Is that better when using the IoU between the prediction box and gt box in v5 experiment？
